File: EmojiPanel/src/main/java/com/binlee/emoji/compat/DefaultHttpEngine.java
package com.binlee.emoji.compat;

import androidx.annotation.NonNull;

import com.binlee.emoji.helper.LogHelper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.Proxy;
import java.net.URL;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public final class DefaultHttpEngine extends HttpEngine<HttpURLConnection, HttpURLConnection> {

    private final Config mConfig;
    private volatile ExecutorService mService;
    private volatile boolean mReauthenticating = false;
    private volatile int redirectCount;

    public DefaultHttpEngine() {
        mConfig = createConfig();
    }

    private ExecutorService service() {
        if (mService == null) {
            mService = new ThreadPoolExecutor(0, Integer.MAX_VALUE, 60, TimeUnit.SECONDS,
                    new SynchronousQueue<>(), r -> new Thread(r, "DefaultHttpEngine"));
        }
        return mService;
    }

    @Override
    protected void setupClient(Object client) {
        if (!(client instanceof HttpURLConnection)) {
            return;
        }
        final HttpURLConnection conn = (HttpURLConnection) client;
        conn.setConnectTimeout(mConfig.connectTimeout * 1000);
        conn.setReadTimeout(mConfig.readTimeout * 1000);
        conn.setInstanceFollowRedirects(true);
        conn.setDoInput(true);
        conn.setUseCaches(false);
    }

    @Override
    public BaseResponse request(@NonNull BaseRequest request) throws IOException {
        if (mReauthenticating) {
            synchronized (mConfig) {
                try {
                    mConfig.wait();
                } catch (InterruptedException ignored) {
                }
            }
        }
        final HttpURLConnection raw = adaptRequest(request);
        writeOutputStream(raw, request);
        return adaptResponse(raw, request);
    }

    @Override
    public void request(@NonNull BaseRequest request, @NonNull Callback callback) {
        // 异步请求要在子线程中执行，然后回调
        service().execute(() -> {
            try {
                final BaseResponse raw = request(request);
                mainHandler().post(() -> callback.onResponse(raw));
            } catch (IOException e) {
                mainHandler().post(() -> callback.onFailure(e));
            }
        });
    }

    @Override
    protected HttpURLConnection adaptRequest(BaseRequest request) throws IOException {
        return createConnection(request);
    }

    @Override
    protected BaseResponse adaptResponse(HttpURLConnection raw, BaseRequest current) throws IOException {
        return resolveConnection(raw, current);
    }

    private HttpURLConnection createConnection(BaseRequest request) throws IOException {
        final String url = buildUrl(request);
        final HttpURLConnection conn = openConnection(url, mConfig.proxy);
        setupClient(conn);
        conn.setRequestMethod(request.method);
        Map<String, String> headers = request.headers == null
                ? new HashMap<>()
                : request.headers;
        if (mConfig.userAgent != null) {
            headers.put("User-Agent", mConfig.userAgent);
        }
        for (String name : headers.keySet()) {
            conn.addRequestProperty(name, headers.get(name));
        }
        return conn;
    }

    private HttpURLConnection openConnection(String url, Proxy proxy) throws IOException {
        if (url.startsWith("https")) {
            // config https
        }
        if (proxy == null) {
            return (HttpURLConnection) new URL(url).openConnection();
        }
        return (HttpURLConnection) new URL(url).openConnection();
    }

    private void writeOutputStream(HttpURLConnection conn, BaseRequest request) throws IOException {
        conn.connect();
        if (!POST.equals(request.method) && !PUT.equals(request.method)) {
            return;
        }
        if (request.params == null || request.params.size() == 0) {
            return;
        }
        // 发送表单数据
        // final BufferedOutputStream out = new BufferedOutputStream(conn.getOutputStream());
    }

    private BaseResponse resolveConnection(HttpURLConnection conn, BaseRequest current)
            throws IOException {
        final BaseResponse response = new BaseResponse();
        // status line
        final String statusLine = conn.getHeaderField(null);
        if (statusLine != null) {
            final String[] elements = statusLine.split(" ");
            if (elements.length > 2) {
                response.protocol = elements[0];
                response.code = Integer.parseInt(elements[1]);
                response.msg = statusLine.substring(statusLine.indexOf(elements[1]) + 4);
            }
        } else {
            response.protocol = conn.getURL().getProtocol();
            response.code = conn.getResponseCode();
            response.msg = conn.getResponseMessage();
        }
        response.headers = new HashMap<>();
        final Map<String, List<String>> headerFields = conn.getHeaderFields();
        // skip status line
        headerFields.remove(null);
        for (String name : headerFields.keySet()) {
            response.headers.put(name, toString(headerFields.get(name)));
        }
        response.current = current;
        if (response.redirect()) {
            final String redirectUrl = response.headers.get("Location");
            if (redirectUrl != null) {
                final int old = redirectCount;
                redirectCount = old + 1;
                LogHelper.debug("DefHttpEngine", "redirect " + redirectUrl + " " + redirectCount);
                final BaseRequest redirect = new BaseRequest();
                redirect.method = current.method;
                redirect.url = redirectUrl;
                redirect.headers = current.headers;
                redirect.params = current.params;
                return request(redirect);
            }
        }
        if (response.clientError()) {
            if (response.code == 401 && mConfig.authenticate != null) {
                // 重新认证
                mReauthenticating = true;
                final BaseResponse raw = request(mConfig.authenticate);
                onReauthenticateResponse(raw);
                synchronized (mConfig) {
                    mReauthenticating = false;
                    mConfig.notifyAll();
                }
                // 继续发送`发生重定向时`的请求
                return request(response.current);
            }
        }
        response.data = toBytes(response.success() ? conn.getInputStream() : conn.getErrorStream());
        if (response.data == null || response.data.length == 0) {
            response.data = response.msg.getBytes();
        }
        trace("DefHttpEngine", response);
        return response;
    }

    private void onReauthenticateResponse(BaseResponse response) {
        LogHelper.debug("DefHttpEngine", "onReauthenticateResponse() " + response);
    }

    private byte[] toBytes(InputStream in) throws IOException {
        // do not close the `in`
        if (in == null) {
            return new byte[0];
        }
        final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        int len;
        byte[] data = new byte[8192];
        while ((len = in.read(data)) != -1) {
            buffer.write(data, 0, len);
        }
        buffer.flush();
        data = buffer.toByteArray();
        buffer.close();
        return data;
    }

    private static String toString(List<String> list) {
        if (list == null) {
            return "";
        }
        final StringBuilder builder = new StringBuilder();
        for (int i = 0; i < list.size(); i++) {
            if (i != 0) {
                builder.append("; ");
            }
            builder.append(list.get(i));
        }
        return builder.toString();
    }
}

File: EmojiPanel/src/main/java/com/binlee/emoji/compat/DefaultHttpEngine.kt
package com.binlee.emoji.compat

import com.binlee.emoji.helper.LogHelper
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.Proxy
import java.net.URL
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.SynchronousQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class DefaultHttpEngine : HttpEngine<HttpURLConnection, HttpURLConnection>() {

    private val mLock = Object()
    private val mConfig: Config?

    @Volatile
    private var mService: ExecutorService? = null

    @Volatile
    private var mReauthenticating = false

    @Volatile
    private var redirectCount = 0
    private fun service(): ExecutorService {
        if (mService == null) {
            mService = ThreadPoolExecutor(0, Int.MAX_VALUE, 60, TimeUnit.SECONDS,
                    SynchronousQueue()) { r: Runnable? -> Thread(r, "DefaultHttpEngine") }
        }
        return mService!!
    }

    override fun setupClient(client: Any?) {
        if (client !is HttpURLConnection) {
            return
        }
        client.connectTimeout = mConfig!!.connectTimeout * 1000
        client.readTimeout = mConfig.readTimeout * 1000
        client.instanceFollowRedirects = true
        client.doInput = true
        client.useCaches = false
    }

    @Throws(IOException::class)
    override fun request(request: BaseRequest): BaseResponse {
        while (mReauthenticating) {
            synchronized(mLock) {
                try {
                    mLock.wait()
                } catch (ignored: InterruptedException) {
                }
            }
        }
        val raw = adaptRequest(request)
        writeOutputStream(raw, request)
        return adaptResponse(raw, request)
    }

    override fun request(request: BaseRequest, callback: Callback) {
        // 异步请求要在子线程中执行，然后回调
        service().execute {
            try {
                val raw = request(request)
                mainHandler().post { callback.onResponse(raw) }
            } catch (e: IOException) {
                mainHandler().post { callback.onFailure(e) }
            }
        }
    }

    @Throws(IOException::class)
    override fun adaptRequest(request: BaseRequest): HttpURLConnection {
        val conn = openConnection(buildUrl(request), mConfig!!.proxy)
        setupClient(conn)
        conn.requestMethod = request.method
        request.headers["User-Agent"] = mConfig.userAgent
        for (name in request.headers.keys) {
            conn.addRequestProperty(name, request.headers[name])
        }
        return conn
    }

    @Throws(IOException::class)
    override fun adaptResponse(raw: HttpURLConnection, current: BaseRequest): BaseResponse {
        return resolveConnection(raw, current)
    }

    @Throws(IOException::class)
    private fun openConnection(url: String?, proxy: Proxy?): HttpURLConnection {
        if (url!!.startsWith("https")) {
            // config https
        }
        return if (proxy == null) {
            URL(url).openConnection() as HttpURLConnection
        } else URL(url).openConnection() as HttpURLConnection
    }

    @Throws(IOException::class)
    private fun writeOutputStream(conn: HttpURLConnection, request: BaseRequest) {
        conn.connect()
        if (POST != request.method && PUT != request.method) {
            return
        }
        if (request.params == null || request.params!!.isEmpty()) {
            return
        }
        // 发送表单数据
        // final BufferedOutputStream out = new BufferedOutputStream(conn.getOutputStream());
    }

    @Throws(IOException::class)
    private fun resolveConnection(conn: HttpURLConnection, current: BaseRequest): BaseResponse {
        val response = BaseResponse()
        // status line
        val statusLine = conn.getHeaderField(null)
        if (statusLine != null) {
            val elements = statusLine.split(" ".toRegex()).toTypedArray()
            if (elements.size > 2) {
                response.protocol = elements[0]
                response.code = elements[1].toInt()
                response.msg = statusLine.substring(statusLine.indexOf(elements[1]) + 4)
            }
        } else {
            response.protocol = conn.url.protocol
            response.code = conn.responseCode
            response.msg = conn.responseMessage
        }
        response.headers = mutableMapOf()
        val headerFields = conn.headerFields
        for (name in headerFields.keys) {
            // skip status line
            if (name != null) {
                response.headers[name] = toString(headerFields[name])
            }
        }
        response.current = current
        if (response.redirect()) {
            val redirectUrl = response.headers["Location"]
            if (redirectUrl != null) {
                val old = redirectCount
                redirectCount = old + 1
                LogHelper.debug("DefHttpEngine", "redirect $redirectUrl $redirectCount")
                val redirect = BaseRequest()
                redirect.method = current.method
                redirect.url = redirectUrl
                redirect.headers = current.headers
                redirect.params = current.params
                return request(redirect)
            }
        }
        if (response.clientError()) {
            if (response.code == 401 && mConfig!!.authenticate != null) {
                // 重新认证
                mReauthenticating = true
                val raw = request(mConfig.authenticate!!)
                onReauthenticateResponse(raw)
                synchronized(mLock) {
                    mReauthenticating = false
                    mLock.notifyAll()
                }
                // 继续发送`发生重新认证`时的请求
                return request(response.current!!)
            }
        }
        response.data = toBytes(if (response.success()) conn.inputStream else conn.errorStream)
        if (response.data == null || response.data!!.isEmpty()) {
            response.data = response.msg!!.toByteArray()
        }
        trace("DefHttpEngine", response)
        return response
    }

    private fun onReauthenticateResponse(response: BaseResponse?) {
        LogHelper.debug("DefHttpEngine", "onReauthenticateResponse() $response")
    }

    @Throws(IOException::class)
    private fun toBytes(`in`: InputStream?): ByteArray {
        // do not close the `in`
        if (`in` == null) {
            return ByteArray(0)
        }
        val buffer = ByteArrayOutputStream()
        var len: Int
        var data = ByteArray(8192)
        while (`in`.read(data).also { len = it } != -1) {
            buffer.write(data, 0, len)
        }
        buffer.flush()
        data = buffer.toByteArray()
        buffer.close()
        return data
    }

    companion object {
        private fun toString(list: List<String>?): String {
            if (list == null) {
                return ""
            }
            val builder = StringBuilder()
            for (i in list.indices) {
                if (i != 0) {
                    builder.append("; ")
                }
                builder.append(list[i])
            }
            return builder.toString()
        }
    }

    init {
        mConfig = createConfig()
    }
}